### PR TITLE
Introduce Linux domain to demonstrate pvnet and pvblock work

### PR DESCRIPTION
**Pull request by oleksiimoisieiev**
This patches are introducing additional domain configuration based on the Linux Kernel with initrd included.
This domain designed to demonstrate pvnet and pvblock initialization.

### CODE REVIEW

File: Kconfig [hunk 1]
```diff
@@ -63,6 +63,14 @@ config DOM_STORAGE_FATFS_LOGICAL_DRIVE_NAME
 	help
 	  FATFS logical drive name to be used as mount point.
 
+config PV_DOMAIN
+	bool "Enable PV domain support loaded from FATFS storage"
+	default y
```
Comment (GrygiriiS) [inline Kconfig]: no default "y" please. enable it explicitly for RPI5

File: Kconfig [hunk 2]
```diff
@@ -63,6 +63,14 @@ config DOM_STORAGE_FATFS_LOGICAL_DRIVE_NAME
 	help
 	  FATFS logical drive name to be used as mount point.
 
+config PV_DOMAIN
```
Comment (GrygiriiS) [inline Kconfig]: Could we name it DOM_CFG_LINUX_PV_DOMAIN?

**Review by GrygiriiS — COMMENTED** [review]
Could you add run example section in README please?

File: src/dom_cfg/rpi_5.c [hunk 3]
```diff
@@ -148,5 +180,11 @@ struct dom0_domain_cfg domain_cfgs[] = {
 		.image_dt_path = DISK_BIN_PATH"z_sync.dtb"
 #endif /* CONFIG_DOM_STORAGE_FATFS_ENABLE */
 	},
+#if defined(CONFIG_PV_DOMAIN)
+	{
+		.domain_cfg = &domu_cfg_2,
+		.image_kernel_path = DISK_BIN_PATH"Image",
```
Comment (GrygiriiS) [inline src/dom_cfg/rpi_5.c]: Please, use better file name

File: src/dom_cfg/rpi_5.c [hunk 4]
```diff
@@ -41,8 +41,40 @@ static ssize_t get_domu_image_size(void *image_info, uint64_t *size)
...
+void fill_pv_params(void) {
+	parse_one_record_and_fill_cfg(params_vif, &domu_cfg_2.back_cfg);
+	parse_one_record_and_fill_cfg(params_vbd, &domu_cfg_2.back_cfg);
+}
+
+#endif /* CONFIG_PV_DOMAIN */
+
 static struct xen_domain_cfg domu_cfg_1 = {
-	.name = "rpi_5_domu",
+	.name = "rpi_5_domu1",
```
Comment (GrygiriiS) [inline src/dom_cfg/rpi_5.c]: no. do not touch it.

File: src/dom_cfg/rpi_5.c [hunk 5]
```diff
@@ -41,8 +41,40 @@ static ssize_t get_domu_image_size(void *image_info, uint64_t *size)
...
+static const char *params_vif =
+	"vif=[ 'backend=1,bridge=xenbr0,mac=08:00:27:ff:cb:ce,ip=192.168.0.2 "
+	"255.255.255.0 192.168.0.1' ]";
+
+static const char *params_vbd =
+	"disk=[ 'backend=1, format=raw, vdev=xvda, "
+                                "access=rw, target=/etc/u-boot-initial-env' ]";
+
+static struct xen_domain_cfg domu_cfg_2 = {
+	.name = "rpi_5_domu2",
```
Comment (GrygiriiS) [inline src/dom_cfg/rpi_5.c]: give it normal informative name - linux_pv_domu, or smth

File: Kconfig (hunk 2 above)
Comment (oleksiimoisieiev) [reply Kconfig]: renamed

File: Kconfig (hunk 1 above)
Comment (oleksiimoisieiev) [reply Kconfig]: +

File: src/dom_cfg/rpi_5.c (hunk 5 above)
Comment (oleksiimoisieiev) [reply src/dom_cfg/rpi_5.c]: +

File: src/dom_cfg/rpi_5.c (hunk 4 above)
Comment (oleksiimoisieiev) [reply src/dom_cfg/rpi_5.c]: +

File: src/dom_cfg/rpi_5.c (hunk 3 above)
Comment (oleksiimoisieiev) [reply src/dom_cfg/rpi_5.c]: +

**Comment by oleksiimoisieiev** [comment]
> Could you add run example section in README please?

I have updated readme

File: README.rst [hunk 6]
```diff
@@ -392,3 +392,58 @@ Attach to DomU console with ``xu console``, use ``Ctrl-']'`` to exit console:
...
+    ~ # ls -la /dev/xvda
+
+Check PV net connection to the Domain:
+
+.. code-block:: console
+    ~ # ip addr add dev eth0 192.168.0.2/24
+    ~ # ip link set eth0 up
+    ~ # ping -c 3 192.168.0.1
+
+**NOTE:** -c 3 parameter for ping is needed because Ctrl+C is not working.
```
Comment (GrygiriiS) [inline README.rst]: .. note::

File: README.rst [hunk 7]
```diff
@@ -392,3 +392,58 @@ Attach to DomU console with ``xu console``, use ``Ctrl-']'`` to exit console:
     thread_a: Hello World from cpu 0 on xenvm!
     thread_b: Hello World from cpu 0 on xenvm!
     thread_a: Hello World from cpu 0 on xenvm!
+
+Linux pv domain control
+=======================
+
+Create linux-pv-domu by using ``xu create`` with linux_pv_domu configuration and Xen
+domainId ``-d 3``.
```
Comment (GrygiriiS) [inline README.rst:400]: no need for "-d 3". I hope https://github.com/xen-troops/zephyr-xenlib/pull/89 will be merged it will be possible to see domid.

File: README.rst [hunk 8]
```diff
@@ -392,3 +392,58 @@ Attach to DomU console with ``xu console``, use ``Ctrl-']'`` to exit console:
...
+    I: kernsize_aligned = 2097152
+    I: DTB will be placed on addr = 0x40e00000
+    I: Created domain is paused
+    ...
+
+Attach to DomU console with ``xu console``, use ``Ctrl-']'`` to exit console:
+
+.. code-block:: console
+
+    uart:~$ xu console 2
```
Comment (GrygiriiS) [inline README.rst]: you do "-d 3 " but here domid=2

File: boards/rpi_5.conf [hunk 9]
```diff
@@ -21,6 +21,9 @@ CONFIG_FILE_SYSTEM_SHELL=y
 
 CONFIG_NOCACHE_MEMORY=y
 CONFIG_FS_MULTI_PARTITION=y
-CONFIG_DOM_STORAGE_FATFS_LOGICAL_DRIVE_NAME="0"
-CONFIG_DOM_STORAGE_FATFS_DIR="dom0/"
+CONFIG_DOM_STORAGE_FATFS_LOGICAL_DRIVE_NAME="1"
```
Comment (GrygiriiS) [inline boards/rpi_5.conf]: This is NOT related to this PR as you switch to dom0 partition, but default ENABLE_DOM0_PART=no

File: src/dom_cfg/rpi_5.c [hunk 10]
```diff
@@ -176,5 +208,11 @@ struct dom0_domain_cfg domain_cfgs[] = {
 		.image_kernel_path = DISK_BIN_PATH"helloworld_xen-arm64",
 	},
 #endif /* CONFIG_DOM_STORAGE_FATFS_ENABLE */
+#if defined(CONFIG_DOM_CFG_LINUX_PV_DOMAIN)
+	{
+		.domain_cfg = &domu_cfg_2,
```
Comment (GrygiriiS) [inline src/dom_cfg/rpi_5.c]: domu_cfg_3!

File: src/dom_cfg/rpi_5.c [hunk 11]
```diff
@@ -63,6 +63,38 @@ static ssize_t get_domu_image_size(void *image_info, uint64_t *size)
...
+#if defined(CONFIG_DOM_CFG_LINUX_PV_DOMAIN)
+static const char *params_vif =
+	"vif=[ 'backend=1,bridge=xenbr0,mac=08:00:27:ff:cb:ce,ip=192.168.0.2 "
+	"255.255.255.0 192.168.0.1' ]";
+
+static const char *params_vbd =
+	"disk=[ 'backend=1, format=raw, vdev=xvda, "
+                                "access=rw, target=/etc/u-boot-initial-env' ]";
+
+static struct xen_domain_cfg domu_cfg_2 = {
```
Comment (GrygiriiS) [inline src/dom_cfg/rpi_5.c]: domu_cfg_3!

File: src/dom0.c [hunk 12]
```diff
@@ -44,7 +48,16 @@ int main(void)
...
 	}
 
-	/* It's required to init xenlib struct xen_domain_cfg->image_info with pointer
+	/*
+	 * Start initialization routine for the pv_domain to parse
+	 * vbd and vif configuration strings and fillup backed parameter
+	 * data.
+	 */
+	if (IS_ENABLED(CONFIG_DOM_CFG_LINUX_PV_DOMAIN)) {
+		fill_pv_params();
```
Comment (GrygiriiS) [inline src/dom0.c]: I'd prefer if you add `.init() `callback in `struct dom0_domain_cfg, `
which accepts `struct dom0_domain_cfg *dom0_cfg` as parameter

Comment (oleksiimoisieiev) [reply src/dom0.c]: This will affect too many entities, such as zephyr-xenlib, xrun tests and so on. I will leave todo here.

Comment (oleksiimoisieiev) [reply src/dom0.c]: I think I've found solution. please ignore previous message

File: README.rst (hunk 7 above)
Comment (oleksiimoisieiev) [reply README.rst:400]: not sure why we need this changes. Currently domid can be seen using xstat stat command. Setting predefined domain id saving me from the necessity of describing how can I get this id. So why not? This command is valid.

File: README.rst (hunk 8 above)
Comment (oleksiimoisieiev) [reply README.rst]: missed that. sry

File: README.rst (hunk 6 above)
Comment (oleksiimoisieiev) [reply README.rst]: +

File: boards/rpi_5.conf (hunk 9 above)
Comment (oleksiimoisieiev) [reply boards/rpi_5.conf]: sry. Could you elaborate?

File: src/dom0.c (hunk 12 above)
Comment (oleksiimoisieiev) [reply src/dom0.c]: fixed

**Comment by GrygiriiS** [comment]
`Reviewed-by: Grygorii Strashko <grygorii_strashko@epam.com>`

File: src/dom_cfg/rpi_5.c [hunk 13]
```diff
@@ -63,6 +63,38 @@ static ssize_t get_domu_image_size(void *image_info, uint64_t *size)
...
 #endif /* CONFIG_DOM_STORAGE_FATFS_ENABLE */
 
+#if defined(CONFIG_DOM_CFG_LINUX_PV_DOMAIN)
+static const char *params_vif =
+	"vif=[ 'backend=1,bridge=xenbr0,mac=08:00:27:ff:cb:ce,ip=192.168.0.2 "
+	"255.255.255.0 192.168.0.1' ]";
+
+static const char *params_vbd =
+	"disk=[ 'backend=1, format=raw, vdev=xvda, "
+                                "access=rw, target=/etc/u-boot-initial-env' ]";
```
Comment (firscity) [inline src/dom_cfg/rpi_5.c]: Strange formatting

**Review by firscity — COMMENTED** [review]
With comment fixed
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`

**Comment by dsemenets** [comment]
`Acked-by: Dmytro Semenets <dmytro_semenets@epam.com>`

File: src/dom_cfg/rpi_5.c (hunk 13 above)
Comment (oleksiimoisieiev) [reply src/dom_cfg/rpi_5.c]: fixed
